File: jane_street/trading_bot/cpp/src/strategy.cpp
#include "strategy.hpp"
#include <algorithm>

namespace trading_bot {

Strategy::Strategy() {
    limits_["BOND"] = 100;
    limits_["VALBZ"] = 10;
    limits_["VALE"] = 10;
    // ... other limits
}

void Strategy::on_hello(const std::vector<Position>& initial_positions) {
    for (const auto& pos : initial_positions) {
        positions_[pos.symbol] = pos.size;
    }
}

void Strategy::on_fill(const FillUpdate& fill) {
    int delta = (fill.dir == Direction::BUY) ? fill.size : -fill.size;
    positions_[fill.symbol] += delta;
}

std::vector<Order> Strategy::decide(const BookUpdate& book) {
    if (book.symbol == "BOND") {
        return bond_strategy(book);
    }
    return {};
}

int Strategy::next_order_id() {
    return ++order_id_counter_;
}

std::vector<Order> Strategy::bond_strategy(const BookUpdate& book) {
    std::vector<Order> orders;
    std::string symbol = "BOND";
    int pos = positions_[symbol];
    int limit = limits_[symbol];

    // Aggressive Buy
    for (const auto& [price, size] : book.sell) {
        if (price < 1000) {
            int buy_size = std::min(size, limit - pos);
            if (buy_size > 0) {
                orders.push_back({next_order_id(), symbol, Direction::BUY, price, buy_size});
                pos += buy_size;
            }
        }
    }

    // Aggressive Sell
    for (const auto& [price, size] : book.buy) {
        if (price > 1000) {
            int sell_size = std::min(size, limit + pos);
            if (sell_size > 0) {
                orders.push_back({next_order_id(), symbol, Direction::SELL, price, sell_size});
                pos -= sell_size;
            }
        }
    }

    // Passive
    if (pos < limit) {
        orders.push_back({next_order_id(), symbol, Direction::BUY, 999, limit - pos});
    }
    if (pos > -limit) {
        orders.push_back({next_order_id(), symbol, Direction::SELL, 1001, limit + pos});
    }

    return orders;
}

} // namespace trading_bot

File: jane_street/trading_bot/rust/src/strategy.rs
use crate::models::{BookUpdate, Direction, FillUpdate, Position, ClientMessage};
use std::collections::HashMap;

pub struct Strategy {
    positions: HashMap<String, i32>,
    order_id_counter: i32,
    limits: HashMap<String, i32>,
}

impl Strategy {
    pub fn new() -> Self {
        let mut limits = HashMap::new();
        limits.insert("BOND".to_string(), 100);
        
        Self {
            positions: HashMap::new(),
            order_id_counter: 0,
            limits,
        }
    }

    pub fn next_order_id(&mut self) -> i32 {
        self.order_id_counter += 1;
        self.order_id_counter
    }

    pub fn on_hello(&mut self, symbols: Vec<Position>) {
        for s in symbols {
            self.positions.insert(s.symbol, s.position);
        }
    }

    pub fn on_fill(&mut self, fill: FillUpdate) {
        let pos = self.positions.entry(fill.symbol).or_insert(0);
        let delta = if fill.dir == Direction::Buy { fill.size } else { -fill.size };
        *pos += delta;
    }

    pub fn decide(&mut self, book: BookUpdate) -> Vec<ClientMessage> {
        if book.symbol == "BOND" {
            return self.bond_strategy(book);
        }
        vec![]
    }

    fn bond_strategy(&mut self, book: BookUpdate) -> Vec<ClientMessage> {
        let mut messages = Vec::new();
        let symbol = "BOND".to_string();
        let mut pos = *self.positions.get(&symbol).unwrap_or(&0);
        let limit = *self.limits.get(&symbol).unwrap_or(&100);

        // Aggressive Buy
        for (price, size) in book.sell {
            if price < 1000 {
                let buy_size = std::cmp::min(size, limit - pos);
                if buy_size > 0 {
                    messages.push(ClientMessage::Add {
                        order_id: self.next_order_id(),
                        symbol: symbol.clone(),
                        dir: Direction::Buy,
                        price,
                        size: buy_size,
                    });
                    pos += buy_size;
                }
            }
        }

        // Aggressive Sell
        for (price, size) in book.buy {
            if price > 1000 {
                let sell_size = std::cmp::min(size, limit + pos);
                if sell_size > 0 {
                    messages.push(ClientMessage::Add {
                        order_id: self.next_order_id(),
                        symbol: symbol.clone(),
                        dir: Direction::Sell,
                        price,
                        size: sell_size,
                    });
                    pos -= sell_size;
                }
            }
        }

        // Passive Buy
        if pos < limit {
            messages.push(ClientMessage::Add {
                order_id: self.next_order_id(),
                symbol: symbol.clone(),
                dir: Direction::Buy,
                price: 999,
                size: limit - pos,
            });
        }

        // Passive Sell
        if pos > -limit {
            messages.push(ClientMessage::Add {
                order_id: self.next_order_id(),
                symbol: symbol.clone(),
                dir: Direction::Sell,
                price: 1001,
                size: limit + pos,
            });
        }

        messages
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::models::Direction;

    #[test]
    fn test_bond_strategy_passive() {
        let mut strategy = Strategy::new();
        strategy.positions.insert("BOND".to_string(), 0);
        let book = BookUpdate {
            symbol: "BOND".to_string(),
            buy: vec![],
            sell: vec![],
        };
        let messages = strategy.decide(book);
        assert!(messages.iter().any(|m| matches!(m, ClientMessage::Add { price: 999, dir: Direction::Buy, .. })));
        assert!(messages.iter().any(|m| matches!(m, ClientMessage::Add { price: 1001, dir: Direction::Sell, .. })));
    }
}
